core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -949,9 +953,65 @@ public void onSuccess() {
+                                                ShardRouting primaryShard = indexShardReference.routingEntry();
+                                                String message = String.format(Locale.ROOT, "primary shard [%s] was demoted while failing replica shard [%s] for [%s]", primaryShard, shard, exp);
+                                                shardStateAction.shardFailed(primaryShard, primaryShard, message, shardFailedError, new ShardStateAction.Listener() {
+                                                    @Override
+                                                    public void onSuccess() {
+                                                        if (clusterService.state().version() == clusterStateVersion) {
+                                                            new ClusterStateObserver(clusterService, null, logger, threadPool.getThreadContext()).waitForNextChange(new ClusterStateObserver.Listener() {
+                                                                @Override
+                                                                public void onNewClusterState(ClusterState state) {
+                                                                    // nothing to do, we just needed a new cluster state
+                                                                    if (logger.isTraceEnabled()) {
+                                                                        logger.trace("previous cluster state version [{}], new cluster state [{}]", clusterStateVersion, state.prettyPrint());
+                                                                    }
+                                                                }
+
+                                                                @Override
+                                                                public void onClusterServiceClose() {
+                                                                    logger.warn("[{}] node [{}] shutdown while waiting to fail demoted primary shard [{}] subsequent to [{}] while failing replica shard [{}]", shardId, clusterService.localNode(), primaryShard, exp, shard);
+                                                                    forceFinishAsFailed(new NodeClosedException(clusterService.localNode()));
+                                                                }
+
+                                                                @Override
+                                                                public void onTimeout(TimeValue timeout) {
+                                                                    // we wait indefinitely
+                                                                }
+                                                            });
+                                                        }
+                                                        new ReroutePhase(null, request, new ActionListener<Response>() {
```

Comment:
I think we can do simpler by just returned a retryable exception to the reroute phase that started all of this. It will do the same thing. Also - I miss the annoying "request" reset (we should open an issue to remove it). I guess it's still coming..